packages/itwin/tree-widget/src/tree-widget-react/components/trees/models-tree/Utils.ts

```diff
@@ -26,3 +26,45 @@ export function releaseMainThreadOnItemsCount<T>(elementCount: number) {
     );
   };
 }
+
+/** @internal */
+export type ChildrenTree<T> = Map<string, { children?: ChildrenTree<T>; additionalProps?: T }>;
+
+/** @internal */
+export function getIdsFromChildrenTree<T>({
+  tree,
+  additionalCheck,
+}: {
+  tree: ChildrenTree<T>;
+  additionalCheck?: (treeEntry: T | undefined) => boolean;
+}): Set<string> {
+  const result = new Set<string>();
+  tree.forEach((entry, id) => {
+    if (!additionalCheck || additionalCheck(entry.additionalProps)) {
+      result.add(id);
+    }
+    if (entry.children) {
+      const childrenIds = getIdsFromChildrenTree({ tree: entry.children, additionalCheck });
+      childrenIds.forEach((childId) => result.add(childId));
+    }
+  });
+  return result;
+}
+
+/** @internal */
+export function getChildIdsFromChildrenTree<T>({
```

Comment:
Not a big fan of having this "almost duplicate" function. Maybe make `additionalCheck` of `getIdsFromChildrenTree` also get a `depth` argument, and then callers could do `getIdsFromChildrenTree(tree, ({ depth }) => depth > 0)`?